File: bst.js
/**
 * Binary Search Tree — Pure JavaScript (ES Module)
 * Self-contained BST class with interactive CLI-style demo via console.
 * Operations: insert, search, delete, traversals, height, count, levelOrder
 */

// ── Node ──────────────────────────────────────────────────────────────────
class BSTNode {
  /** @param {number} value */
  constructor(value) {
    this.value = value;
    /** @type {BSTNode|null} */ this.left  = null;
    /** @type {BSTNode|null} */ this.right = null;
  }
}

// ── BST Class ─────────────────────────────────────────────────────────────
export class BST {
  constructor() {
    /** @type {BSTNode|null} */
    this.root = null;
  }

  // ── Insert ───────────────────────────────────────────────────────────
  /** @param {number} value */
  insert(value) {
    this.root = this.#insertRec(this.root, value);
    return this; // fluent API
  }

  /** @param {BSTNode|null} node @param {number} value @returns {BSTNode} */
  #insertRec(node, value) {
    if (!node) return new BSTNode(value);
    if (value < node.value)      node.left  = this.#insertRec(node.left,  value);
    else if (value > node.value) node.right = this.#insertRec(node.right, value);
    // duplicates are silently ignored
    return node;
  }

  // ── Search ───────────────────────────────────────────────────────────
  /** @param {number} value @returns {boolean} */
  search(value) {
    return this.#searchRec(this.root, value);
  }

  #searchRec(node, value) {
    if (!node) return false;
    if (value === node.value) return true;
    return value < node.value
      ? this.#searchRec(node.left,  value)
      : this.#searchRec(node.right, value);
  }

  // ── Delete ───────────────────────────────────────────────────────────
  /** @param {number} value */
  delete(value) {
    this.root = this.#deleteRec(this.root, value);
    return this;
  }

  #deleteRec(node, value) {
    if (!node) return null;
    if (value < node.value) {
      node.left  = this.#deleteRec(node.left,  value);
    } else if (value > node.value) {
      node.right = this.#deleteRec(node.right, value);
    } else {
      // Node found — three cases
      if (!node.left)  return node.right;
      if (!node.right) return node.left;
      // Two children: replace with inorder successor
      const successor = this.#minNode(node.right);
      node.value = successor.value;
      node.right = this.#deleteRec(node.right, successor.value);
    }
    return node;
  }

  #minNode(node) {
    let cur = node;
    while (cur.left) cur = cur.left;
    return cur;
  }

  // ── Height & Count ───────────────────────────────────────────────────
  height() { return this.#heightRec(this.root); }
  #heightRec(n) { return n ? 1 + Math.max(this.#heightRec(n.left), this.#heightRec(n.right)) : 0; }

  count() { return this.#countRec(this.root); }
  #countRec(n) { return n ? 1 + this.#countRec(n.left) + this.#countRec(n.right) : 0; }

  get isEmpty() { return this.root === null; }

  // ── Traversals ───────────────────────────────────────────────────────
  /** @returns {number[]} Values in ascending order */
  inorder() {
    const result = [];
    const walk = (n) => { if (!n) return; walk(n.left); result.push(n.value); walk(n.right); };
    walk(this.root);
    return result;
  }

  /** @returns {number[]} Root first */
  preorder() {
    const result = [];
    const walk = (n) => { if (!n) return; result.push(n.value); walk(n.left); walk(n.right); };
    walk(this.root);
    return result;
  }

  /** @returns {number[]} Root last */
  postorder() {
    const result = [];
    const walk = (n) => { if (!n) return; walk(n.left); walk(n.right); result.push(n.value); };
    walk(this.root);
    return result;
  }

  /** @returns {number[][]} Each sub-array is one level */
  levelOrder() {
    if (!this.root) return [];
    const result = [];
    const queue  = [this.root];
    while (queue.length) {
      const level = [];
      const size  = queue.length;
      for (let i = 0; i < size; i++) {
        const node = queue.shift();
        level.push(node.value);
        if (node.left)  queue.push(node.left);
        if (node.right) queue.push(node.right);
      }
      result.push(level);
    }
    return result;
  }

  // ── ASCII Print ──────────────────────────────────────────────────────
  /** Renders a sideways ASCII tree to console */
  print() {
    const lines = [];
    const walk = (n, prefix, isLeft) => {
      if (!n) return;
      walk(n.right, prefix + (isLeft ? '│   ' : '    '), false);
      lines.push(prefix + (isLeft ? '└── ' : '┌── ') + n.value);
      walk(n.left,  prefix + (isLeft ? '    ' : '│   '), true);
    };
    if (this.root) { walk(this.root.right, '', false); lines.push(String(this.root.value)); walk(this.root.left, '', true); }
    console.log(lines.join('\n'));
  }
}

// ── Demo (runs when executed directly with Node.js) ───────────────────────
const tree = new BST();
const values = [50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45];

console.log('=== Binary Search Tree — JavaScript ===\n');
console.log('Inserting:', values.join(', '));
values.forEach(v => tree.insert(v));

console.log('\nASCII Tree:');
tree.print();

console.log('\nInorder:   ', tree.inorder().join(' → '));
console.log('Preorder:  ', tree.preorder().join(' → '));
console.log('Postorder: ', tree.postorder().join(' → '));
console.log('LevelOrder:', tree.levelOrder().map((l,i) => `L${i}:[${l}]`).join('  '));

console.log('\nHeight:    ', tree.height());
console.log('Count:     ', tree.count());

console.log('\nSearch 40: ', tree.search(40));
console.log('Search 99: ', tree.search(99));

console.log('\nDeleting 30 (two children)...');
tree.delete(30);
console.log('Inorder:   ', tree.inorder().join(' → '));

console.log('Deleting 80 (leaf)...');
tree.delete(80);
console.log('Inorder:   ', tree.inorder().join(' → '));

console.log('\nFinal count:', tree.count(), '| Height:', tree.height());

File: BST.java
// Binary Search Tree — Java Implementation
// Operations: insert, search, delete, inorder, preorder, postorder traversal

public class BST {

    static class Node {
        int value;
        Node left, right;

        Node(int value) {
            this.value = value;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    public BST() {
        root = null;
    }

    // ── Insert ─────────────────────────────────────────────────────────────
    public void insert(int value) {
        root = insertRec(root, value);
    }

    private Node insertRec(Node node, int value) {
        if (node == null) return new Node(value);
        if (value < node.value)      node.left  = insertRec(node.left,  value);
        else if (value > node.value) node.right = insertRec(node.right, value);
        // duplicate values are ignored
        return node;
    }

    // ── Search ─────────────────────────────────────────────────────────────
    public boolean search(int value) {
        return searchRec(root, value);
    }

    private boolean searchRec(Node node, int value) {
        if (node == null) return false;
        if (value == node.value) return true;
        return value < node.value
            ? searchRec(node.left,  value)
            : searchRec(node.right, value);
    }

    // ── Delete ─────────────────────────────────────────────────────────────
    public void delete(int value) {
        root = deleteRec(root, value);
    }

    private Node deleteRec(Node node, int value) {
        if (node == null) return null;

        if (value < node.value) {
            node.left = deleteRec(node.left, value);
        } else if (value > node.value) {
            node.right = deleteRec(node.right, value);
        } else {
            // Node to delete found
            if (node.left == null)  return node.right;
            if (node.right == null) return node.left;

            // Node has two children — replace with inorder successor
            node.value = minValue(node.right);
            node.right = deleteRec(node.right, node.value);
        }
        return node;
    }

    private int minValue(Node node) {
        int min = node.value;
        while (node.left != null) {
            min  = node.left.value;
            node = node.left;
        }
        return min;
    }

    // ── Height ─────────────────────────────────────────────────────────────
    public int height() {
        return heightRec(root);
    }

    private int heightRec(Node node) {
        if (node == null) return 0;
        return 1 + Math.max(heightRec(node.left), heightRec(node.right));
    }

    // ── Traversals ─────────────────────────────────────────────────────────
    public void inorder() {
        System.out.print("Inorder:   ");
        inorderRec(root);
        System.out.println();
    }

    private void inorderRec(Node node) {
        if (node == null) return;
        inorderRec(node.left);
        System.out.print(node.value + " ");
        inorderRec(node.right);
    }

    public void preorder() {
        System.out.print("Preorder:  ");
        preorderRec(root);
        System.out.println();
    }

    private void preorderRec(Node node) {
        if (node == null) return;
        System.out.print(node.value + " ");
        preorderRec(node.left);
        preorderRec(node.right);
    }

    public void postorder() {
        System.out.print("Postorder: ");
        postorderRec(root);
        System.out.println();
    }

    private void postorderRec(Node node) {
        if (node == null) return;
        postorderRec(node.left);
        postorderRec(node.right);
        System.out.print(node.value + " ");
    }

    // ── Level-Order (BFS) ──────────────────────────────────────────────────
    public void levelOrder() {
        if (root == null) { System.out.println("Tree is empty."); return; }
        java.util.Queue<Node> queue = new java.util.LinkedList<>();
        queue.add(root);
        System.out.print("LevelOrder: ");
        while (!queue.isEmpty()) {
            Node current = queue.poll();
            System.out.print(current.value + " ");
            if (current.left  != null) queue.add(current.left);
            if (current.right != null) queue.add(current.right);
        }
        System.out.println();
    }

    // ── Count Nodes ────────────────────────────────────────────────────────
    public int countNodes() {
        return countRec(root);
    }

    private int countRec(Node node) {
        if (node == null) return 0;
        return 1 + countRec(node.left) + countRec(node.right);
    }

    // ── Demo ───────────────────────────────────────────────────────────────
    public static void main(String[] args) {
        BST tree = new BST();
        int[] values = {50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45};

        System.out.println("=== Binary Search Tree — Java ===\n");
        System.out.print("Inserting: ");
        for (int v : values) {
            System.out.print(v + " ");
            tree.insert(v);
        }
        System.out.println("\n");

        tree.inorder();
        tree.preorder();
        tree.postorder();
        tree.levelOrder();

        System.out.println("\nHeight:     " + tree.height());
        System.out.println("Node count: " + tree.countNodes());

        System.out.println("\nSearch 40:  " + tree.search(40));
        System.out.println("Search 99:  " + tree.search(99));

        System.out.println("\nDeleting 30 (node with two children)...");
        tree.delete(30);
        tree.inorder();

        System.out.println("Deleting 20 (node with one child)...");
        tree.delete(20);
        tree.inorder();

        System.out.println("Deleting 10 (leaf node)...");
        tree.delete(10);
        tree.inorder();

        System.out.println("\nFinal node count: " + tree.countNodes());
    }
}
